### PR TITLE
MAID-1957 test/tunnel: tunnel client reconnect

### PR DESCRIPTION
**Pull request by maqi**

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @afck

(maidsafe_highfive has picked a reviewer for you, use r? to override)

**Comment by maqi** [comment]
> we could modify the mock crust framework to enable the blocking/dropping of tunnelled connections there, and just allow the production code to take care of handling the generated crust events
> when we remove a connection we can check for a NodeLost event

File: tests/mock_crust/tunnel.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        node.inner.set_next_node_name(name1);
+    }
+    nodes.push(TestNode::builder(network).config(config.clone()).create());
+
+    (Endpoint(nodes.len() - 2), Endpoint(nodes.len() - 1))
+}
+
+#[test]
+fn tunnel_clients() {
+    let min_section_size = 5;
```
Comment (afck) [inline tests/mock_crust/tunnel.rs]: Maybe we could even use size 3, to speed up the test?

File: tests/mock_crust/tunnel.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        .rev()
+        .filter_map(|(index, node)| if node.routing_table().our_prefix().matches(prefix_name) {
+            Some(index)
+        } else {
+            None
+        })
+        .collect();
+    section_indexes.iter()
+        .take(section_indexes.len() - min_section_size + 1)
+        .foreach(|index| { let _ = nodes.remove(*index); });
```
Comment (dhardy) [inline tests/mock_crust/tunnel.rs]: You'd best sort the indices from highest to lowest if you want to remove them from the `Vec` like this.

File: tests/mock_crust/tunnel.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    poll_and_resend(nodes, &mut []);
+}
+
+// Adds a pair of nodes with specified names into the network. Returns the endpoints of the nodes.
```
Comment (dhardy) [inline tests/mock_crust/tunnel.rs]: Also blocks direct connection between these two nodes if `is_tunnel` is set.

File: tests/mock_crust/utils.rs [hunk 4]
```diff
@@ -378,26 +390,26 @@ pub fn create_connected_nodes_with_cache(network: &Network, size: usize, use_cac
 //
 // The array is sanity checked (e.g. it would be an error to pass [1, 1, 1]), must comprise at
 // least two elements, and every element must be no more than `8`.
-pub fn create_connected_nodes_until_split(network: &Network,
-                                          mut prefix_lengths: Vec<usize>,
-                                          use_cache: bool)
-                                          -> Nodes {
+pub fn create_connected_nodes_until_split_with_nodes(network: &Network,
```
Comment (dhardy) [inline tests/mock_crust/utils.rs]: Should be `add_connected_nodes_until_split`.

File: tests/mock_crust/tunnel.rs (hunk 2 above)
Comment (maqi) [inline tests/mock_crust/tunnel.rs]: it was taken in a reversed order, https://github.com/maidsafe/routing/pull/1355/files#diff-b83639c52df93eae4d3e66ec59a8e66fR69
so I think we are ok here?

Comment (dhardy) [inline tests/mock_crust/tunnel.rs]: I guess so. I don't much like this style of coding because it's easy to accidentally break it, but I guess it'll do.

**Pull request merged**